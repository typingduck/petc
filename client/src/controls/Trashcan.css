
#petc-trashcan {
  bottom: 10px;
  color: rgba(0,0,0,0.1);
  font-size: 96px;
  left: 50%;
  margin-left: -50px; 
  position: absolute;
  text-align: center;
  width: 100px;
  z-index: var(--z-index-page-bg)
}

#petc-trashcan.active {
  color: black;
}

#petc-trashcan.delete {
  color: red;
}

#petc-trashcan.hidden {
  /* Make transparent instead of hide so that react bounding rectangle work */
  color: rgba(0, 0, 0, 0.0);
}
