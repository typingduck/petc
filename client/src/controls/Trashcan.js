import React from 'react'
import ReactDOM from 'react-dom'
import {MdDelete} from 'react-icons/lib/md'

import './Trashcan.css'

export function isInTrashCan (node) {
  return bounds && withinRect(node, bounds)
}

let bounds = null

class Trashcan extends React.Component {
  componentDidMount () {
    /* eslint-disable react/no-find-dom-node */
    let dom = ReactDOM.findDOMNode(this)
    if (dom) {
      bounds = dom.getBoundingClientRect()
    }
    /* eslint-enable react/no-find-dom-node */
  }

  render () {
    let className = 'hidden'

    if (this.props.controls.isNodeMode) {
      className = ''
      const dragNode = this.props.controls.dragNode
      if (dragNode) {
        if (isInTrashCan(dragNode)) {
          className = 'delete'
        } else {
          className = 'active'
        }
      }
    }
    return (
      <div id='petc-trashcan' className={className}>
        <MdDelete />
      </div>
    )
  }
}

function within (v, a, b) {
  return v >= a && v <= b
}

function withinRect (node, rect) {
  return within(node.x, rect.left, rect.right) &&
      within(node.y, rect.top, rect.bottom)
}

export default Trashcan
