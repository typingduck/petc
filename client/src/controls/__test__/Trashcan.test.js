/* global expect it jest */
import React from 'react'
import Renderer from 'react-test-renderer'

import Trashcan from '../Trashcan'

jest.mock('react-dom', () => ({
  findDOMNode: () => ({
    getBoundingClientRect: () => ({
      left: 10,
      right: 20,
      top: 100,
      bottom: 110
    })
  })
}))

it('hidden when not editing nodes', () => {
  const props = {
    controls: {
      isNodeMode: false,
      dragNode: { id: 'id', x: 15, y: 105 }
    }
  }
  const component = Renderer.create(<Trashcan {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})

it('visible when editing nodes', () => {
  const props = {
    controls: {
      isNodeMode: true,
      dragNode: null
    }
  }
  const component = Renderer.create(<Trashcan {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})

it('active when editing nodes and have dragging node', () => {
  const props = {
    controls: {
      isNodeMode: true,
      dragNode: { id: 'id', x: 1, y: 1 }
    }
  }
  const component = Renderer.create(<Trashcan {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})

it('red when node inside trashcan', () => {
  const props = {
    controls: {
      isNodeMode: true,
      dragNode: { id: 'id', x: 15, y: 105 }
    }
  }
  const component = Renderer.create(<Trashcan {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})
