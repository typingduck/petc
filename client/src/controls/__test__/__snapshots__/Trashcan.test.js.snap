// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`active when editing nodes and have dragging node 1`] = `
<div
  className="active"
  id="petc-trashcan"
>
  <svg
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "color": undefined,
        "verticalAlign": "middle",
      }
    }
    viewBox="0 0 40 40"
    width="1em"
  >
    <g>
      <path
        d="m31.6 6.6v3.4h-23.2v-3.4h5.7l1.8-1.6h8.2l1.8 1.6h5.7z m-21.6 25v-20h20v20c0 1.8-1.6 3.4-3.4 3.4h-13.2c-1.8 0-3.4-1.6-3.4-3.4z"
      />
    </g>
  </svg>
</div>
`;

exports[`hidden when not editing nodes 1`] = `
<div
  className="hidden"
  id="petc-trashcan"
>
  <svg
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "color": undefined,
        "verticalAlign": "middle",
      }
    }
    viewBox="0 0 40 40"
    width="1em"
  >
    <g>
      <path
        d="m31.6 6.6v3.4h-23.2v-3.4h5.7l1.8-1.6h8.2l1.8 1.6h5.7z m-21.6 25v-20h20v20c0 1.8-1.6 3.4-3.4 3.4h-13.2c-1.8 0-3.4-1.6-3.4-3.4z"
      />
    </g>
  </svg>
</div>
`;

exports[`red when node inside trashcan 1`] = `
<div
  className="delete"
  id="petc-trashcan"
>
  <svg
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "color": undefined,
        "verticalAlign": "middle",
      }
    }
    viewBox="0 0 40 40"
    width="1em"
  >
    <g>
      <path
        d="m31.6 6.6v3.4h-23.2v-3.4h5.7l1.8-1.6h8.2l1.8 1.6h5.7z m-21.6 25v-20h20v20c0 1.8-1.6 3.4-3.4 3.4h-13.2c-1.8 0-3.4-1.6-3.4-3.4z"
      />
    </g>
  </svg>
</div>
`;

exports[`visible when editing nodes 1`] = `
<div
  className=""
  id="petc-trashcan"
>
  <svg
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "color": undefined,
        "verticalAlign": "middle",
      }
    }
    viewBox="0 0 40 40"
    width="1em"
  >
    <g>
      <path
        d="m31.6 6.6v3.4h-23.2v-3.4h5.7l1.8-1.6h8.2l1.8 1.6h5.7z m-21.6 25v-20h20v20c0 1.8-1.6 3.4-3.4 3.4h-13.2c-1.8 0-3.4-1.6-3.4-3.4z"
      />
    </g>
  </svg>
</div>
`;
